model/project_event.go

```diff
@@ -22,14 +22,86 @@ type ProjectSettings struct {
 	Subscriptions      []event.Subscription `bson:"subscriptions" json:"subscriptions"`
 }
 
+type ProjectSettingsEvent struct {
+	ProjectSettings              `bson:",inline"`
+	FilesIgnoredFromCacheDefault bool `bson:"files_ignored_from_cache_default,omitempty" json:"files_ignored_from_cache_default,omitempty"`
+	GitTagAuthorizedTeamsDefault bool `bson:"git_tag_authorized_teams_default,omitempty" json:"git_tag_authorized_teams_default,omitempty"`
+	GitTagAuthorizedUsersDefault bool `bson:"git_tag_authorized_users_default,omitempty" json:"git_tag_authorized_users_default,omitempty"`
+	PatchTriggerAliasesDefault   bool `bson:"patch_trigger_aliases_default,omitempty" json:"patch_trigger_aliases_default,omitempty"`
+	PeriodicBuildsDefault        bool `bson:"periodic_builds_default,omitempty" json:"periodic_builds_default,omitempty"`
+	TriggersDefault              bool `bson:"triggers_default,omitempty" json:"triggers_default,omitempty"`
+	WorkstationCommandsDefault   bool `bson:"workstation_commands_default,omitempty" json:"workstation_commands_default,omitempty"`
+}
+
 type ProjectChangeEvent struct {
-	User   string          `bson:"user" json:"user"`
-	Before ProjectSettings `bson:"before" json:"before"`
-	After  ProjectSettings `bson:"after" json:"after"`
+	User   string               `bson:"user" json:"user"`
+	Before ProjectSettingsEvent `bson:"before" json:"before"`
+	After  ProjectSettingsEvent `bson:"after" json:"after"`
 }
 
 type ProjectChangeEvents []ProjectChangeEventEntry
 
+// Look at the flags set in the ProjectSettingsEvent so that fields that were converted to empty arrays when casting to an interface{} can be correctly set to nil
```

Comment:
I'm probably missing some context, but couldn't we just iterate and check for empty arrays, and convert all empty arrays to nil?

Reply:
Is there ever a time when we specifically want there to be an empty array? I would assume that an empty array could always be safely changed to nil?

Reply:
The use case is a setting like this, where "override" means empty array and "default" means null
<img width="410" alt="image" src="https://user-images.githubusercontent.com/9298431/186916449-421595c6-0fdf-4506-acf8-ac2f464bda47.png">

All of the settings I added flags for use this pattern, which is why assuming null on empty array is invalid.

Reply:
nit: function comments start with the function name and explain what the function does before going into details